### PR TITLE
Cs/sc 3360 support road network algorithm v2

### PR DESCRIPTION
**Pull request by Charl1996**
Same as https://github.com/dimagi/commcare-hq/pull/34024, but with commit history cleaned up a bit.

## Product Description
[Demo video](https://vimeo.com/905617454/0f7b334709?share=copy) (note that the FF was removed)

The disbursement algorithm, "Road Network Algorithm",  makes use of the Mapbox Matrix API. This PR adds the ability for a project to add their own mapbox token on the Geospatial Settings page. 

image

Once the FF is enabled for a project, the new disbursement algorithm will appear on the Geospatial Settings page. Once the algorithm is selected, two new HTML elements will appear on the screen:
1) Text (password) input prompting the user for their mapbox token.
2) A button ("Test API Key") which the user can click to make a sample request to Mapbox in order to verify that the token works.

Note that the user don't have to click the "Test API Key" button in order to save the token; the button exists purely to test the token. 

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3360)

Noteworthy items:
1) The api token is stored in postgres as an encrypted string, but returned to the front end in plaintext (although rendered as a password) when loading the Settings page.
2) The "Test API Key" button sample API call happens purely in the browser; I decided we don't have to have a dedicated endpoint just for the server to make an API call to Mapbox if we could simply use ajax.  

## Feature Flag
Geospatial FF

### Safety story
Tested this feature on my local machine.

### Automated test coverage
- Tests for view
- Tests for model

### QA Plan
QA planned

### Migrations
- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/geospatial/models.py [hunk 1]
```diff
@@ -81,3 +83,22 @@ def disbursement_solver(self):
...
+            ciphertext = self.api_token.split('$', 2)[2]
+            return b64_aes_decrypt(ciphertext)
+        return self.api_token
+
+    @plaintext_api_token.setter
+    def plaintext_api_token(self, value):
+        if value is None:
+            self.api_token = None
+        elif not isinstance(value, str):
+            raise Exception("Only string values allowed for plaintext api token")
```
Comment (SmittieC) [inline corehq/apps/geospatial/models.py:99]: How is this propegated to the user? Do we catch it and show a message to the user? If so, we should translate it as well

**Review by SmittieC — COMMENTED** [review]
Just one comment, but looks fine to me

**Comment by mkangia** [comment]
Good to avoid adding-removing the feature flag if possible for even better commit history.
I assume you could simply remove the first commit and edit the one that removes the FF.

**Review by zandre-eng — COMMENTED** [review]
Took a pass and overall looks good, given that there are no new code changes from the previous PR.

Happy to approve once outstanding comment from Chris is addressed.

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 2]
```diff
@@ -107,7 +106,7 @@ def calculate_distance_matrix(self):
             'sources': sources,
             'destinations': destinations,
             'annotations': 'distance',
-            'access_token': settings.MAPBOX_ACCESS_TOKEN
+            'access_token': config.plaintext_api_token,
```
Comment (mkangia) [inline corehq/apps/geospatial/routing_solvers/pulp.py:109]: Will this impact the projects using radial algorithm? they don't need to have a token right? so we should use the one in settings for them?

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 3]
```diff
@@ -46,6 +50,32 @@ hqDefine("geospatial/js/geo_config", [
...
 
+        self.selectedAlgorithm = ko.observable();
+        self.plaintext_api_token = ko.observable(data.plaintext_api_token);
+
+        self.captureApiToken = ko.computed(function () {
+            return self.selectedAlgorithm() === ROAD_NETWORK_ALGORITHM;
+        });
+
+        self.validateApiToken = function () {
+            const url = "https://api.mapbox.com/directions-matrix/v1/mapbox/driving/-122.42,37.78;-122.45,37.91;-122.48,37.73";
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/geo_config.js:61]: just confirming that mapbox does not provide an endpoint just to validate the token?

File: corehq/apps/geospatial/models.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/geospatial/models.py:99]: This isn't being propagated, as the input from the user is always string. I just thought I'd add this as an additional safeguard.

File: corehq/apps/geospatial/tests/test_models.py [hunk 4]
```diff
@@ -21,6 +21,37 @@ def test_geo_config(self):
...
+            config.plaintext_api_token = '1234'
+            self.assertEqual(config.plaintext_api_token, '1234')
+            self.assertTrue(config.api_token.startswith(f"${ALGO_AES}$"))
+
+            config.plaintext_api_token = None
+            self.assertEqual(config.plaintext_api_token, None)
+            self.assertEqual(config.api_token, None)
+
+    def test_geo_config_api_token_cannot_be_non_str(self):
+        with self.assertRaises(Exception) as context:
```
Comment (mkangia) [inline corehq/apps/geospatial/tests/test_models.py:35]: is `assertRaisesWithMessage` not available here?

File: corehq/apps/geospatial/tests/test_models.py [hunk 5]
```diff
@@ -21,6 +21,37 @@ def test_geo_config(self):
...
+        self.assertEqual(str(context.exception), "Only string values allowed for plaintext api token")
+
+    def test_geo_config_api_token_cannot_be_empty(self):
+        with self.assertRaises(Exception) as context:
+            with self.get_geo_config() as config:
+                config.plaintext_api_token = ""
+
+        self.assertEqual(str(context.exception), "Unexpected value set for plaintext api token")
+
+    def test_geo_config_api_token_cannot_start_with_encryption_str(self):
```
Comment (mkangia) [inline corehq/apps/geospatial/tests/test_models.py:48]: nice one!

File: corehq/apps/geospatial/tests/test_views.py [hunk 6]
```diff
@@ -162,6 +162,45 @@ def test_config_update(self):
         self.assertEqual(config.target_group_count, 10)
         self.assertEqual(config.selected_disbursement_algorithm, GeoConfig.RADIAL_ALGORITHM)
 
+    @flag_enabled('GEOSPATIAL')
+    def test_config_update__road_network_algorithm_ff_disabled(self):
```
Comment (mkangia) [inline corehq/apps/geospatial/tests/test_views.py:166]: the name says "ff_disabled", I assume it was when we added a new feature flag but that is now removed?

File: submodules/langcodes
Comment (mkangia) [inline submodules/langcodes:1]: is this change expected?

**Review by mkangia — CHANGES_REQUESTED** [review]
Thanks for a better PR @Charl1996 
Just one blocking feedback to check on and some minor queries.

File: corehq/apps/geospatial/tests/test_views.py (hunk 6 above)
Comment (Charl1996) [reply corehq/apps/geospatial/tests/test_views.py:166]: Ah, sharp eyes! I'll update

File: corehq/apps/geospatial/routing_solvers/pulp.py (hunk 2 above)
Comment (Charl1996) [reply corehq/apps/geospatial/routing_solvers/pulp.py:109]: > Will this impact the projects using radial algorithm?
The `RoadNetworkSolver` class inherits from `RadialDistanceSolver`, but the `RadialDistanceSolver` class has it's own `calculate_distance_matrix` method which does not make use of mapbox.

File: corehq/apps/geospatial/models.py (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/models.py:99]: In the very unlikely event that this exception gets thrown, would it then result in an uncaught exception? If so, I wonder if it might be better to at least safely catch this from where we're calling this?

Comment (Charl1996) [reply corehq/apps/geospatial/models.py:99]: I do agree it might be prudent to catch the exception, but I'm not 100% sure what to do with it, unless I do something like the following...
```
    def save(self, commit=True):
        if self.cleaned_data.get('plaintext_api_token'):
            try:
                self.instance.plaintext_api_token = self.cleaned_data.get('plaintext_api_token')
            except ValueError:
                pass
        return super().save(commit)
```
This will simply not save the erroneous token, but without telling the user why.

The form already handles the case where a token is an empty string, but not ints or bools...since the user can't set a faulty value from the UI, I don't think there's much sense in trying to report such a case to the user.  

Thoughts?

Comment (SmittieC) [reply corehq/apps/geospatial/models.py:99]: I see the form specifies this field to be a CharField, so all characters will be interpreted as string already, making this check redundant?

Comment (zandre-eng) [reply corehq/apps/geospatial/models.py:99]: Instead of doing a type check here, what if we handled this in the `clean()` function of the form? We can check there whether the type is a string and, if not, raise an appropriate `ValidationError`. This way, we can be sure that the data coming from the front-end is a string and return an appropriate error if not.

File: submodules/langcodes
Comment (mkangia) [reply submodules/langcodes:1]: that came from sub module.
It seems you used `git add .`
good to edit the commit that added this.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 3 above)
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:61]: @mkangia No, I'm not aware of any such endpoint.

Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:61]: Cool, thanks @Charl1996

**Comment by Charl1996** [comment]
Closing once again in favour of https://github.com/dimagi/commcare-hq/pull/34114

File: submodules/langcodes
Comment (Charl1996) [reply submodules/langcodes:1]: See [this PR](https://github.com/dimagi/commcare-hq/pull/34114).
